File: Python/QuickSort.py
def partition(Array: list, beg: int, end: int) -> int:
    pivot = Array[end]
    i = beg - 1
    for j in range(beg, end):
        if (Array[j] <= pivot):
            i += 1
            # exchange(Array[i], Array[j])
            temp = Array[i]
            Array[i] = Array[j]
            Array[j] = temp

    # exchange(Array[i + 1], Array[end])
    Array[i + 1], Array[end] = Array[end], Array[i + 1]
    return i + 1


def quickSort(Array: list, beg: int, end: int) -> None:
    if (beg < end):
        # print(Array)
        pivot = partition(Array, beg, end)
        quickSort(Array, beg, (pivot - 1))
        quickSort(Array, (pivot + 1), end)


if __name__ == '__main__':
    Array = [7, 8, 16, 1, 2, 5]
    print('Unsorted Array:')
    print(Array)
    quickSort(Array, 0, len(Array) - 1)
    print('Unsorted Array:')
    print(Array)

File: C/quicksort.c
#include <stdio.h>
#include <stdlib.h>

int partition(int a[], int p, int r)
{
    int x, j, i, t;
    x = a[r];
    i = p - 1;
    for (j = p; p < r; j++)
    {
        if (a[j] <= x)
        {
            i = i + 1;

            t = a[i];
            a[i] = a[j];
            a[j] = t;
        
        }
        
        t = a[i + 1];
        a[i + 1] = a[r];
        a[r] = t;
        
        return a[i + 1];
    }
}
void quicksort(int a[], int p, int r)
{
    int q;
    if (p < r)
    {
        q = partition(a, p, r);
        quicksort(a, p, q - 1);
        quicksort(a, q + 1, r);
    }
}
void main()
{
    int a[] = {3, 2, 4, 5, 1, 0};
    quicksort(a, 0, 5);
    for (int i = 0; i < 5; i++)
    {
        printf("%d", a[i]);
    }
}
